tests/equation_bcs/test_bcs.py

```diff
@@ -0,0 +1,64 @@
+import pytest
+
+from firedrake import *
+
+
+@pytest.mark.parallel
+def test_bc_on_sub_sub_domain():
+
+    # Solve a vector poisson problem
+
+    mesh = UnitSquareMesh(32, 32)
+
+    V = VectorFunctionSpace(mesh, "CG", 4)
+    VV = MixedFunctionSpace([V, V])
+
+    x, y = SpatialCoordinate(mesh)
+
+    f = Function(V)
+    f.interpolate(as_vector([-8.0 * pi * pi * cos(x * pi * 2) * cos(y * pi * 2),
+                             -8.0 * pi * pi * cos(x * pi * 2) * cos(y * pi * 2)]))
+
+    # interpolate the exact solution on gg[i][j]
+    gg = [[None, None], [None, None]]
+    for i in [0, 1]:
+        for j in [0, 1]:
+            gg[i][j] = Function(VV.sub(i).sub(j))
+            gg[i][j].interpolate(cos(2 * pi * x) * cos(2 * pi * y))
+
+    uu = Function(VV)
+    vv = TestFunction(VV)
+
+    F = 0
+    for u, v in zip(split(uu), split(vv)):
+        F += (- inner(grad(u), grad(v)) - dot(f, v)) * dx
+
+    bcs = [DirichletBC(VV.sub(0).sub(0), gg[0][0], 1),
+           DirichletBC(VV.sub(0).sub(1), gg[0][1], 2),
+           DirichletBC(VV.sub(1).sub(0), gg[1][0], 3),
+           DirichletBC(VV.sub(1).sub(1), gg[1][1], "on_boundary")]
+
+    parameters = {
+        "mat_type": "matfree",
+        "snes_max_it": 1,
+        "ksp_type": "gmres",
+        "ksp_rtol": 1.e-10,
+        "ksp_atol": 1.e-10,
+        "ksp_max_it": 200000,
+        "pc_type": "fieldsplit",
+        "pc_fieldsplit_type": "schur",
+        "pc_fieldsplit_schur_fact_type": "full",
+        "fieldsplit_0_ksp_type": "gmres",
+        "fieldsplit_0_ksp_rtol": 1.e-12,
+        "fieldsplit_1_ksp_type": "gmres",
+        "fieldsplit_1_ksp_rtol": 1.e-12,
```

Comment:
To avoid having to do millions of iterations, do:
```python
parameters = {
  "mat_type": "matfree",
  "snes_type": "ksponly",
  "ksp_type": "preonly",
  "pc_type": "fieldsplit",
  "pc_fieldsplit_type": "additive",
  "fieldsplit_ksp_type": "preonly",
  "fieldsplit_pc_type": "python",
  "fieldsplit_pc_python_type": "firedrake.AssembledPC",
  "fieldsplit_assembled_pc_type": "lu"
}
```

I think

Reply:
Thanks!